### ISSUE TITLE
Type mismatch not detected for inheriting classes

### ISSUE THREAD

**Issue by harsimranmaan** [Microsoft/TypeScript#2860]
```javascript
class Player{}
class Batter extends Player{}
class Bowler extends Player{}
class Game{
    private _batter:Batter;
    private _bowler:Bowler;
    constructor( batter:Batter,  bowler:Bowler)
    {
       this._batter = batter;
       this._bowler = bowler; 
    }
}
// Type mismatch for Batter and Bowler, pay attention to the order and the Game constructor
var game = new Game(new Bowler(), new Batter());
var game = new Game(new Batter(), new Bowler());
```

**Comment by DanielRosenwasser** [comment]
@danquirk explains it very nicely [here](https://github.com/Microsoft/TypeScript/issues/2288#issuecomment-78155379). I'll quote him and adapt his response to your example.

> This is by design. TypeScript uses structural typing to determine assignability. The name of a type is irrelevant, what matters is its shape. In this case your *[Batter]* and *[Bowler]* types are identical based on their shape, so they are assignable to one another. If they had a disjoint set of properties you would see the error you expect. This design is intended to enable common JavaScript/TypeScript patterns that mix named types (classes/interfaces) with unnamed types (object literals and function types).

**Comment by DanielRosenwasser** [comment]
Note that as Dan mentioned, you could fix this by adding a disjoint set of properties to get the expected error:

```TypeScript
class Player {
    private __playerTag: Player;
}
class Batter extends Player {
    private __batterTag: Batter;
}
class Bowler extends Player {
    private __bowlerTag: Bowler;
}

```

**Comment by harsimranmaan** [comment]
Makes sense. Thanks :+1:
